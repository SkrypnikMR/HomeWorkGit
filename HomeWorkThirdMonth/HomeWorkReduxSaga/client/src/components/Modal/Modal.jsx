import React, { useState, useEffect } from 'react';
import propTypes from 'prop-types';
import './Modal.scss';
import Button from '../Button';
import Input from '../Input';

const Modal = ({ postMovies, onButtonClick, title: propsTitle, myClassName: propsShow }) => {
    const [state, setState] = useState({
        title: '',
        cover: '',
        description: '',
        show: false,
        inputs: [{ placeholder: 'enter title here', name: 'title' },
        { placeholder: 'enter cover here', name: 'cover' },
        { placeholder: 'enter description here', name: 'description' }],
    });
    useEffect(() => setState({ ...state, show: propsShow }), [propsShow]);

    const handleOnChange = (e) => setState({ ...state, [`${e.target.name}`]: e.target.value });
    const handleClean = () => setState({ ...state, title: '', cover: '', description: '' })
    const handleAddMovie = () => {
        const { title, cover, description } = state;
        const object = { title, cover, description };
        postMovies(object);
        onButtonClick();
        handleClean();
    }
    const handleBtnHeaderClick = (e) => {
        if (e.target.classList.contains('modal') || e.target.classList.contains('custom__button')) {
            onButtonClick();
            handleClean();
        }
    }
    return (
        <div className={state.show ? 'modal' : 'modal hide'} onClick={handleBtnHeaderClick}>
            <div className="modal__content" >
                <div className="content__header">
                    <h3>{propsTitle}</h3>
                    <Button title="X" onButtonClick={handleBtnHeaderClick} />
                </div>
                <div className="content__body">
                    {state.inputs.map((input) => (
                        <Input
                            key={input.name}
                            name={input.name}
                            value={state[input.name]}
                            placeholder={input.placeholder}
                            onChangeInput={handleOnChange}
                        />
                    ))}
                    <Button
                        title="Add movie"
                        clsNm="movie__add"
                        disabled={!state.cover || !state.description || !state.title}
                        onButtonClick={handleAddMovie}
                    />
                </div>
            </div>
        </div>
    );
}
Modal.propTypes = {
    myClassName: propTypes.bool,
    onButtonClick: propTypes.func,
    title: propTypes.string,
    postMovies: propTypes.func,
};
export default Modal;
