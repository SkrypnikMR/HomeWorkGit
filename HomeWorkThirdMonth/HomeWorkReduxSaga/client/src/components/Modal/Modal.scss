
.modal {
    width: 100%;
    height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    position: fixed;
    background-color: rgb(11, 10, 10, 0.5);
    &__content{
        border: 1px solid black;
        border-radius: 15px;
        min-width: 40%;
        height: 700px;
        background-color: #1B264F;
        display: flex;
        justify-content: flex-start;
        align-items: center;
        flex-direction: column;
        .content__header{
            background-color: #f5da0b;
               border-radius: 15px 15px 0 0;
            justify-self: flex-start;
            width: 100%;
            min-height: 50px;
            display: flex;
            justify-content: space-between;
            align-items: center;

            font-size: 20px;
            font-weight: bold;
                h3{ 
                    z-index: 1000;
                    display:flex; 
                    justify-content: center;
                    margin-left:30%;
                }
             button {
              font-weight: bold;
              position: relative;
              top:0px;
              right: 30px;
              padding-right: 10px;
              width: 50px;
              transition-duration: .5s;
              &:hover{
                font-weight: bold;
                transition-duration: .5s;
                font-size: 30px;
                color: #1B264F;
            }
        }
        }
        .content__body{
            margin-top:150px;
            display: flex;
            justify-content: center;
            align-items: center;
            flex-direction: column;
            width: 100%;
            overflow: hidden;
            input{
                min-height: 50px;
                width: 80%;
                margin: 20px;
                font-family: 'Play', sans-serif;
                font-weight: bold;
                color:#f5da0b;
            }
        }
    }
}
.hide {
    display: none;
}
