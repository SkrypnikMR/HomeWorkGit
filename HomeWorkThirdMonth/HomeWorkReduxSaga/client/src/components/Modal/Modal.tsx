import React, { useState, useEffect } from 'react';
import './Modal.scss';
import Button from '../Button';
import Input from '../Input';

interface IModal{
    postMovies: (obj: IpostMovies) => void;
    onButtonClick: () => void;
    title: string;
    myClassName: boolean;
}

interface IpostMovies {
    title: string;
    cover: string;
    description: string;
}

interface IState {
    [key: string]: any;
    title: string;
    cover: string;
    description: string;
    show: boolean;
    inputs: { placeholder: string; name: string }[];
};

const Modal = ({ postMovies, onButtonClick, title: propsTitle, myClassName: propsShow }: IModal) => {
    const [state, setState] = useState<IState>({
        title: '',
        cover: '',
        description: '',
        show: false,
        inputs: [{ placeholder: 'enter title here', name: 'title' },
        { placeholder: 'enter cover here', name: 'cover' },
        { placeholder: 'enter description here', name: 'description' }],
    });
    useEffect(() => setState({ ...state, show: propsShow }), [propsShow]);

    const handleOnChange = (e :React.ChangeEvent<HTMLInputElement> ) => setState({ ...state, [`${e.target.name}`]: e.target.value });
    const handleClean = () => setState({ ...state, title: '', cover: '', description: '' })
    const handleAddMovie = () => {
        const { title, cover, description } = state;
        const object = { title, cover, description };
        postMovies(object);
        onButtonClick();
        handleClean();
    }
    const handleBtnHeaderClick = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
        if (e.target.classList.contains('modal') || e.target.classList.contains('custom__button')) {
            onButtonClick();
            handleClean();
        }
    }
    return (
        <div className={state.show ? 'modal' : 'modal hide'} onClick={handleBtnHeaderClick}>
            <div className="modal__content" >
                <div className="content__header">
                    <h3>{propsTitle}</h3>
                    <Button title="X" onButtonClick={handleBtnHeaderClick} />
                </div>
                <div className="content__body">
                    {state.inputs.map((input) => (
                        <Input
                            key={input.name}
                            name={input.name}
                            value={state[input.name]}
                            placeholder={input.placeholder}
                            onChangeInput={handleOnChange}
                        />
                    ))}
                    <Button
                        title="Add movie"
                        clsNm="movie__add"
                        disabled={!state.cover || !state.description || !state.title}
                        onButtonClick={handleAddMovie}
                    />
                </div>
            </div>
        </div>
    );
}

export default Modal;
