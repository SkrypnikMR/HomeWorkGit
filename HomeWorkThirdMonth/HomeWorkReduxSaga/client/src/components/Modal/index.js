import { connect } from 'react-redux';
import Modal from './Modal';
import { postMovies } from '../../store/movies/actions';

const mapDispatchToProps = { postMovies };

export default connect(null, mapDispatchToProps)(Modal);
