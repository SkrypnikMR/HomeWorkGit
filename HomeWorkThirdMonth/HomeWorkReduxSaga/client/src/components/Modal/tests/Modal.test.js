import React from 'react';
import { shallow, mount } from 'enzyme';
import Modal from '../Modal';

describe('Modal', () => {
    it('Should match snapshot', () => {
        const component = shallow(<Modal />);
        expect(component.html()).toMatchSnapshot();
    });
    it('Should have Modal ', () => {
        const component = mount(<Modal />);
        expect(component.find('Modal')).toHaveLength(1);
    });
    it('Should have divs', () => {
        const component = mount(<Modal />);
        expect(component.find('div')).toHaveLength(4);
    });
    it('Should have divs', () => {
        const component = mount(<Modal />);
        expect(component.find('div').at(0).hasClass('modal hide')).toBe(true);
    });
    it('Should have divs', () => {
        const component = mount(<Modal myClassName={true} />);
        expect(component.find('div').at(0).hasClass('modal')).toBe(true);
        expect(component.find('div').at(0).hasClass('hide')).toBe(false);
    });
    it('Should have div modal__content', () => {
        const component = mount(<Modal />);
        expect(component.find('div').at(1).hasClass('modal__content')).toBe(true);
    });
    it('Should have div content__header', () => {
        const component = mount(<Modal />);
        expect(component.find('div').at(2).hasClass('content__header')).toBe(true);
    });
    it('Should have h3', () => {
        const component = mount(<Modal />);
        expect(component.find('h3')).toHaveLength(1);
    });
    it('Should have div content__body', () => {
        const component = mount(<Modal />);
        expect(component.find('div').at(3).hasClass('content__body')).toBe(true);
    });
    it('Should have div content__body', () => {
        const component = mount(<Modal />);
        expect(component.find('div').at(3).hasClass('content__body')).toBe(true);
    });
    it('Should have Input', () => {
        const component = mount(<Modal />);
        expect(component.find('Input')).toHaveLength(3);
    });
    it('Should change Input 1 ', () => {
        const component = mount(<Modal />);
        component.find('Input').at(0).simulate('change', { target: { name: 'cover', value: 'newCover' } })
    });
    it('Should change Input 2', () => {
        const component = mount(<Modal />);
        component.find('Input').at(1).simulate('change', { target: { name: 'description', value: 'newDescription' } })
    });
    it('Should change Input 3', () => {
        const component = mount(<Modal />);
        component.find('Input').at(2).simulate('change', { target: { name: 'title', value: 'newTitle' } })
    });
    it('Should not disabled addMovie button', () => {
        const component = mount(<Modal />);
        component.find('Input').at(0).simulate('change', { target: { name: 'cover', value: 'newCover' } })
        component.find('Input').at(1).simulate('change', { target: { name: 'description', value: 'newDescription' } })
        component.find('Input').at(2).simulate('change', { target: { name: 'title', value: 'newTitle' } })
        expect(component.find('Button').at(1).props().disabled).toBe(false);
    });
    it('Should click on modal field', () => {
        const postMovies = jest.fn();
        const onButtonClick = jest.fn();
        const component = mount(<Modal postMovies={postMovies} onButtonClick={onButtonClick} />);
        component.find('div').at(0).simulate('click');
        expect(component.find('div').at(0).hasClass('hide')).toBe(true);
    });
    it('Should click addMovie button and call functions', () => {
        const postMovies = jest.fn();
        const onButtonClick = jest.fn();
        const component = mount(<Modal postMovies={postMovies} onButtonClick={onButtonClick} />);
        component.find('Input').at(0).simulate('change', { target: { name: 'cover', value: 'newCover' } })
        component.find('Input').at(1).simulate('change', { target: { name: 'description', value: 'newDescription' } })
        component.find('Input').at(2).simulate('change', { target: { name: 'title', value: 'newTitle' } })
        component.find('Button').at(1).simulate('click');
        expect(postMovies).toHaveBeenCalled();
        expect(onButtonClick).toHaveBeenCalled();
    });
});
