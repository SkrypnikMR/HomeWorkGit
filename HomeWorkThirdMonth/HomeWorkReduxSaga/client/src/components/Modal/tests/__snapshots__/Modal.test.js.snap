// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Should match snapshot 1`] = `"<div class=\\"modal hide\\"><div class=\\"modal__content\\"><div class=\\"content__header\\"><h3></h3><button class=\\"custom__button\\">X</button></div><div class=\\"content__body\\"><input type=\\"text\\" autoComplete=\\"off\\" placeholder=\\"enter title here\\" value=\\"\\" name=\\"title\\"/><input type=\\"text\\" autoComplete=\\"off\\" placeholder=\\"enter cover here\\" value=\\"\\" name=\\"cover\\"/><input type=\\"text\\" autoComplete=\\"off\\" placeholder=\\"enter description here\\" value=\\"\\" name=\\"description\\"/><button disabled=\\"\\" class=\\"movie__add\\">Add movie</button></div></div></div>"`;
